### ISSUE TITLE
Tensor.add_() returns different values on different runs using some inputs (using CPU).

### ISSUE THREAD

**Issue by sakhar** [pytorch/pytorch#14463]
## 🐛 Bug

When using the CPU only (no CUDA), the in-place add function `torch.Tensor.add_()` returns different values on different calls (with trivial differences) using some inputs. However, `torch.Tensor.add()` does not.

## To Reproduce

1. use this numpy array as an example [https://raw.githubusercontent.com/sakhar/pytorch_add_bug/master/values-np.p](https://raw.githubusercontent.com/sakhar/pytorch_add_bug/master/values-np.p)
or [values-np.p.zip](https://github.com/pytorch/pytorch/files/2623714/values-np.p.zip) (unzip).

2. try it with `add_()`:

```python
for i in range(10):
    values = pickle.load(open('values-np.p','rb'))
    x = torch.from_numpy(values[0])
    y = torch.from_numpy(values[1])
    print(x.add_(-0.7,y).data.detach().numpy()[-4])
```
**results:** 
```
using add_():
-0.09407266
-0.094072655
-0.09407266
-0.094072655
-0.09407266
-0.09407266
-0.094072655
-0.094072655
-0.09407266
-0.09407266
```
2. with `add()`:
```python
for i in range(10):
    values = pickle.load(open('values-np.p','rb'))
    x = torch.from_numpy(values[0])
    y = torch.from_numpy(values[1])
    print(x.add(-0.7,y).data.detach().numpy()[-4])
```

**results:**
```
using add():
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
-0.094072655
```

## Environment

```

 - PyTorch Version: 0.4.1
 - OS: Mac OSX 10.12.6
 - How you installed PyTorch: pip
 - Python version: 2.7
 - CUDA/cuDNN version: No CUDA
 - GPU models and configuration: No CUDA

- Versions of relevant libraries:
[pip] numpy (1.14.5)
[pip] torch (0.4.1)
[pip] torchvision (0.2.1)
```

cc @ezyang @gchanan @zou3519

**Comment by vishwakftw** [comment]
The difference observed is in the 7th digit after the decimal point, and is perhaps a floating point issue rather than a functional issue.

**Comment by apaszke** [comment]
While the difference really is very small it is quite suspicious behaviour, because it shouldn't matter if you read the values in place or not. It's not like floating point math is non-deterministic.

**Comment by sakhar** [comment]
I just realized that this behavior does not only depend on a particular element of a tensor, but it depends on the other values in the tensor. Here is an example where I use only the element that its final value (after adding using `add()` changes on different runs:
``` python
for i in range(10):
    values = pickle.load(open('values-np.p','rb'))
    x = torch.from_numpy(values[0][-4:-3])
    y = torch.from_numpy(values[1][-4:-3])
    print(x.add(-0.7,y).data.detach().numpy())
```

**Comment by sakhar** [comment]
> Can you convert the torch tensor to a list before printing, via the `.tolist()`? Just to factor out that this is not a printing issue

The same pattern with different precision.

``` python
for i in range(10):
    values = pickle.load(open('values-np.p','rb'))
    x = torch.from_numpy(values[0])
    y = torch.from_numpy(values[1])
    res = x.add_(-0.7,y)
    print 'to_list():', res[-4].tolist()
    print 'numpy:', res.data.detach().numpy()[-4]
    print
```

result:
```
to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726548433
numpy: -0.094072655

to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726548433
numpy: -0.094072655

to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726548433
numpy: -0.094072655

to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726622939
numpy: -0.09407266

to_list(): -0.0940726622939
numpy: -0.09407266
```

**Comment by sakhar** [comment]
Also, the returned numbers by torch are always inaccurate even when I use `add()`, it gives: `-0.0940726548433`
While the correct result is `x-0.7y = -0.09407265968620777` where:
`x = -0.0723949819803`
`y = 0.0309681110084`

Numpy returns the correct (or more accurate) result.

You can use an external calculator to check that the results returned by numpy is the correct one.

**Comment by albanD** [comment]
Hi,

With the floating point semantic, both results are correct. There isn't one that is more accurate than the other.

Also I cannot reproduce this issue on any of the Linux machines I have, but I can reproduce on a colleague's Mac. Unfortunately I cannot use that machine for further investigation.
The next debugging step I think would be to print the bit representation of both inputs and outputs and see what is the difference.

**Comment by cpuhrsch** [comment]
cc @colesbury

**Comment by cpuhrsch** [comment]
@sakhar - is it at all possible for you to update to a more recent version of pytorch?

**Comment by colesbury** [comment]
In 0.4.1, depending on the CPU architecture, we used [FMA](https://en.wikipedia.org/wiki/Multiply%E2%80%93accumulate_operation) instructions for add(). These are more precise than a multiply followed by an add because it does only one rounding instead of two.

The old implementation used FMA instructions for first N elements where N is a multiple of 8. The remaining elements used a C loop which may not be compiled with FMAs:

https://github.com/pytorch/pytorch/blob/a24163a95edb193ff7b06e98cd69bf7cfd4c0d2f/aten/src/TH/vector/AVX2.cpp#L26-L28

**Comment by apaszke** [comment]
@colesbury all that makes sense. the problem is that the different results are not at the end of the tensor, but are scattered across it.

**Comment by colesbury** [comment]
@apaszke, you're right, that doesn't explain the problem. We were actually using `THBlas(axpy)` for in-place contiguous addition:

https://github.com/pytorch/pytorch/blob/v0.4.1/aten/src/TH/generic/THTensorMath.cpp#L1301-L1303

My guess is that the BLAS implementation on Mac depends on the alignment of the data. For 32-byte aligned data it probably uses FMA and for non-aligned data it probably uses separate multiply and addition instructions. The alignment is non-deterministic and every loop iteration re-allocates the NumPy array.

Anyways, the code path is pretty different in master and I don't think this problem exists there.

**Comment by ja3067** [comment]
Has any work been done on this? This was actually a significant issue while grading homework using PyTorch (in the Columbia University machine learning class). Non-determinism in vanilla gradient descent on Linux (Colab) vs. Mac (most students) machines made grading extremely hard. I feel like basic functionality like addition and gradient descent ought to be deterministic on all machines.

**Comment by albanD** [comment]
@ja3067  This is a different issue. Here we try to get reproducible results for a given machine with a fixed software.

Across different machines, this is absolutely not possible: Addition on different CPUs can already give you different results due to multithreading for large enough Tensors.
Similarly, for different GPUs, you will get different results as they have different properties.

**Comment by ja3067** [comment]
@albanD Thanks for the response. That makes sense. We're also seeing differences on the same machine between torch.add_ and the infix addition and multiplication operator:

```python
for param in net.parameters():
        param.data.add_(param.grad.data, alpha=-eta)
```

```python
for param in net.parameters():
        param -= eta * param.grad
```

return different results. Is this intended. I don't mean to hijack this thread if this is a different issue/not intended to be guaranteed.

**Comment by albanD** [comment]
You can open an post on the forum https://discuss.pytorch.org/ if you have more questions.
But the two above are expected to not be perfectly equal. Floating point operations are not associative on a computer I'm afraid.

**Comment by zou3519** [comment]
We should check if the original issue exists on master or not anymore. It probably doesn't anymore.

I am marking this as hi-pri because this is an easy check (install latest pytorch binary and run code).

**Comment by albanD** [comment]
I tried but I cannot load the repro file anymore on the python 3.7 on my machine.
I'll optimistically close this. Please re-open if you still see it happening on master.
